pylmod/stellargradebook.py

```diff
@@ -0,0 +1,641 @@
+#!/usr/bin/python
+#
+# Python interface to Stellar Grade Book module
+#
+# Defines the class StellarGradeBook
+
+import csv
+import json
+import logging
+import pickle
+import time
+
+from lxml.html.soupparser import fromstring as fromstring_bs
+import requests
+
+
+log = logging.getLogger(__name__)
+
+
+class StellarGradeBook(object):
+    """
+    Python class representing interface to Stellar gradebook.
+
+    Example usage:
+
+    sg = StellarGradeBook('ichuang-cert.pem')
+    ats = sg.get('academicterms')
+    tc = ats['data'][0]['termCode']
+    sg.get('academicterm',termCode=tc)
+
+    students = sg.get_students()
+    assignments = sg.get_assignments()
+    sg.create_assignment('midterm1', 'mid1', 1.0, 100.0, '11-04-2013')
+
+    sid, student = sg.get_student_by_email(email)
+    aid, assignment = sg.get_assignment_by_name('midterm1')
+    sg.set_grade(aid, sid, 95.2)
+
+    sg.spreadsheet2gradebook(datafn)
+
+    """
+
+    URLBASE = 'https://learning-modules.mit.edu:8443/service/gradebook'
+
+    GETS = {'academicterms': '',
+            'academicterm': '/{termCode}',
+            'gradebook': '?uuid={uuid}',
+    }
+
+    GBUUID = 'STELLAR:/project/mitxdemosite'
+    TIMEOUT = 200  # connection timeout, seconds
+
+    verbose = True
+    gradebookid = None
+
+    def __init__(
+        self, cert, urlbase=None, touchstone=False,
+        cookies=None, gbuuid=None
+    ):
+        """
+        Initialize StellarGradeBook instance.
+
+          - urlbase:    URL base for gradebook API (defaults to self.URLBASE)
+          - touchstone: True if touchstone authentication to be done
+            (still needs certs); default False
+          - cookies:    requests.CookieJar object to use (optional)
+          - gbuuid:     gradebook UUID (eg STELLAR:/project/mitxdemosite)
+
+        """
+        # pem with private and public key application certificate for access
+        self.cert = cert
+
+        if urlbase is not None:
+            self.URLBASE = urlbase
+        self.ses = requests.Session()
+        self.ses.cert = cert
+        self.ses.timeout = self.TIMEOUT  # connection timeout
+        self.ses.verify = True  # verify site certificate
+
+        log.debug("------------------------------------------------------")
+        log.info("[StellarGradeBook] init base=%s" % urlbase)
+
+        if cookies is not None:
+            if type(cookies) == str:
+                self.load_cookies_from_file(cookies)
+            else:
+                self.ses.cookies = cookies
+        self.do_touchstone = touchstone
+        if touchstone and cookies is None:
+            self.init_touchstone()
+        if gbuuid is not None:
+            self.gradebookid = self.get_gradebookid(gbuuid)
+
+    def load_cookies_from_file(self, fn):
```

Comment:
We shouldn't need any of this cookie code, see https://trello-attachments.s3.amazonaws.com/53e3c7df2821e27d93802f51/54b7f7074cb6b803f91b1837/32451737a78a71d9776fdb0f9fb1fc7b/cover.html.tar.gz for information on functions that aren't needed.